feat: Add support to create new change event (#78)
newrelic/deployment-marker-action: README.md

A GitHub Action to add New Relic deployment markers during your release pipeline.

> The Action supports two primary API mutations for submitting events:

Author: done.

#### changeTrackingCreateDeployment (Legacy):

> This mutation is designed specifically for submitting traditional deployment events. It represents the older method for sending deployment-focused information to New Relic.

Author: done.

#### changeTrackingCreateEvent (RECOMMENDED):

> This is our latest and most flexible API mutation, offering extensive customization capabilities. It allows you to comprehensively track any type of change event, which includes, but is not limited to, deployments, along with custom attributes, feature flag rollouts, and various custom change events.

Author: done.

It is important to note this will create a `ChangeTrackingEvent` event instead of a `Deployment` event and is a paid feature, billed per event rather than by data volume (per GB).

The `command_type` must be set to `change-tracking` to enable this functionality. Please refer below example for more details and its usage.
For more details on change tracking event mutation, please refer to our [docs](https://docs.newrelic.com/docs/change-tracking/change-tracking-events/#change-tracking-event-mutation)

## Inputs

| Key                    | Required | Default        | Description                                                                                                                                                                                                           |
|------------------------|----------|----------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `guid`                 | yes      | -              | The entity GUID to apply the deployment marker.                                                                                                                                                                       |

Reviewer: I understand that. But you are currently indicating that it's always required and I'm saying it shouldn't be required if they use the new mutation. There are some use cases where using the entity name is much easier for our customers. 

Proposal: `command_type` should actually be `commandType` to fit in with the existing naming convention. It can contain one of two values: `createEvent` or `createDeployment`, with it defaulting to `createDeployment` if nothing is passed for backwards compatibility. 

I think you should have two separate tables, one for each `commandType`.

A GitHub Action to add New Relic deployment markers during your release pipeline.

The Action supports two primary NerdGraph API mutations for submitting events:

> #### changeTrackingCreateDeployment (Legacy):

Reviewer 2: Let's have this one go second.

> This mutation is designed specifically for submitting traditional `Deployment` events. It represents the older method for sending deployment-focused information to New Relic.

Reviewer 2: ```suggestion
This mutation is designed specifically for submitting legacy Deployment events. It represents the older method for sending deployment-focused information to New Relic.

This this will create a `Deployment` event. 
```

> #### changeTrackingCreateEvent (RECOMMENDED):

Reviewer 2: Can we have the `commandType` directly mirror the mutation name -- changeTrackingCreateEvent -- rather than createEvent?

> This is our latest and most flexible API mutation, offering extensive customization capabilities. It allows you to comprehensively track any type of change event, which includes, but is not limited to Deployments, along with custom attributes, feature flag rollouts, customizing the marker and activity stream text via `shortDescription`, and various custom change events.

Reviewer 2: ```suggestion
Our latest API mutation with the most customizations and flexibility. Track any type of change event including Deployments or any other category and type that you would like to send, such as Feature Flag and Business Events. Add custom attributes to decorate your events with any relevant business metadata. Use our entity search to place change events on entities without needing to specify the entity guid. Customize the marker and activity stream text via `shortDescription` so you can see all of the relevant details in the New Relic UI at a glance.
```

> It is important to note this will create a `ChangeTrackingEvent` event instead of a `Deployment` event and is a paid feature, billed per event rather than by data volume (per GB).

Reviewer 2: ```suggestion
This this will create a `ChangeTrackingEvent` event. 
```

The `commandType` must be set to `createEvent` to enable this functionality. Please refer to the example below for more details and its usage.

> For more details on change tracking event mutation, please refer to our [docs](https://docs.newrelic.com/docs/change-tracking/change-tracking-events/#change-tracking-event-mutation)

Reviewer 2: ```suggestion
For more details about the `changeTrackingCreateEvent` mutation, please refer to our [docs](https://docs.newrelic.com/docs/change-tracking/change-tracking-events/#change-tracking-event-mutation)
```

> ## changeTrackingCreateDeployment Inputs

Reviewer 2: Let's swap the order here with the `changeTrackingCreateEvent`.

| Key                    | Required | Default        | Description                                                                                                                                                                                                           |
|------------------------|----------|----------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `guid`                 | yes      | -              | The entity GUID to apply the deployment marker.                                                                                                                                                                       |

Reviewer 2: The ordering is strange here. How about this: Required fields, then optional fields. And within that, sort alpha for each group.

| `user`                 | yes      | `github.actor` | A username to associate with the deployment, visible in the Overview page and on the Deployments page.                                                                                                                |

## changeTrackingCreateEvent Inputs

| Key                    | Required | Default        | Description                                                                                                                                                                                                           |
|------------------------|----------|----------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `apiKey`               | yes      | -              | Your New Relic [personal API key](https://docs.newrelic.com/docs/apis/get-started/intro-apis/types-new-relic-api-keys#personal-api-key).                                                                              |

> | `changelog`            | no       | -              | A summary of what changed in this deployment, visible in the Deployments page.                                                                                                                                        |

Reviewer 2: ```suggestion
| `changelog`            | no       | -              | When category is `Deployment`, a summary of what changed in this deployment, visible in the Change Details page.                                                                                                                                        |
```

> | `commit`               | no       | -              | The Commit SHA for this deployment, visible in the Deployments page.                                                                                                                                                  |

Reviewer 2: ```suggestion
| `commit`               | no       | -              | When category is `Deployment`, the Commit SHA for this deployment, visible in the Change Details page.                                                                                                                                                  |
```

> | `description`          | no       | -              | A high-level description of this deployment, visible in the Overview page and on the Deployments page when you select an individual deployment.                                                                       |

Reviewer 2: ```suggestion
| `description`          | no       | -              | A high-level description of this change event, visible in the Overview page and on the Change Details page when you select an individual change event.                                                                       |
```

> | `deeplink`             | no       | -              | A deep link to the source which triggered the deployment.                                                                                                                                                             |

Reviewer 2: ```suggestion
| `deeplink`             | no       | -              | When category is `Deployment`, a deep link to the source which triggered the deployment.                                                                                                                                                             |
```

> | `deploymenttype`       | no       | `BASIC`        | The type of deployment. Choose from BASIC, BLUE_GREEN, CANARY, OTHER, ROLLING, or SHADOW.                                                                                                                             |

Reviewer 2: Remove, this should be used here.

> | `groupid`              | no       | -              | A group ID for the deployment to link to other deployments.                                                                                                                                                           |

Reviewer 2: ```suggestion
| `groupid`              | no       | -              | A group ID for the change event to link to other change events.                                                                                                                                                           |
```

| `region`               | no       | `US`           | The region of your New Relic account. Default: `US`                                                                                                                                                                   |

> | `version`              | no       | -              | Metadata to apply to the deployment marker - e.g. the latest release tag. Note: This is required if the category is set to "Deployment".                                                                              |

Reviewer 2: ```suggestion
| `version`              | no       | -              | When category is `Deployment`, the version of the deployment - e.g. the latest release tag. Note: This is required if the category is set to "Deployment".                                                                              |
```

> | `user`                 | yes      | `github.actor` | A username to associate with the deployment, visible in the Overview page and on the Deployments page.                                                                                                                |

Reviewer 2: ```suggestion
| `user`                 | yes      | `github.actor` | A username to associate with the deployment, visible in the Overview page and on the Change Details page.                                                                                                                |
```

> | `commandType`          | no       | -              | This field lets you track any change event. To use this new API mutation, simply set it to "createEvent".                                                                                                             |

Reviewer 2: ```suggestion
| `commandType`          | yes       | -              | This field lets you track any change event. To use this new API mutation, you must set it to "createEvent".                                                                                                             |
```

| `category`             | yes      | -              | The category of the change event. For a list of supported categories, [view our docs](https://docs.newrelic.com/docs/change-tracking/change-tracking-events/#supported-types).                                        |
| `type`                 | yes      | -              | The type of the change event. For a list of supported types, [view our docs](https://docs.newrelic.com/docs/change-tracking/change-tracking-events/#supported-types)                                                  |

> | `featureFlagId`        | no       | -              | The ID of the feature flag associated with the change event. This is required if the category is "Feature Flag".                                                                                                      |

Reviewer 2: ```suggestion
| `featureFlagId`        | no       | -              | When category is `Feature Flag`, the ID of the feature flag associated with the change event. This is required if the category is "Feature Flag".                                                                                                      |
```

          deploymenttype: "ROLLING"
          groupId: "Workshop App Release: ${{ github.ref_name }}"
          user: "${{ github.actor }}"

>       # When chaining steps together, the deployment id is placed into the github environment 

Reviewer 2: For the createEvent mutation, we'll have a changeTrackingId. Calling it a `deploymentId` might be confusing.

      - name: View output
        run: echo "${{ env.deploymentId }}"
```

### changeTrackingCreateEvent (RECOMMENDED)

> The below example outlines the usage of the `changeTrackingCreateEvent` capability, will allow deployment events to be reported as rich Change Tracking events, providing more comprehensive metadata and better integration with New Relic's observability platform.

Reviewer 2: ```suggestion
The below example outlines the usage of the `changeTrackingCreateEvent` capability. It will allow change events to be reported as rich Change Tracking events, providing more comprehensive metadata and better integration with New Relic's observability platform.
```

For more details on change tracking event mutation, please refer to our [docs](https://docs.newrelic.com/docs/change-tracking/change-tracking-events/#change-tracking-event-mutation)

```yaml
jobs:
  newrelic:
    runs-on: ubuntu-latest
    name: New Relic ChangeTracking Event 
    steps:
      # This step builds a var with the release tag value to use later
      - name: Set Release Version from Tag
        run: echo "RELEASE_VERSION=${{ github.ref_name }}" >> $GITHUB_ENV
      - name: Test change event deployment marker
        uses: newrelic/deployment-marker-action@v2.2.0
        with:
          apiKey: ${{ secrets.NEW_RELIC_API_KEY }}
          commandType: "createEvent"

>           entitySearch: "id='entity-guid'"

Reviewer 2: How would they pull this from the environment/configuration? What about an example with entity name and domain/type?

Author: done.

          version: "${{ env.RELEASE_VERSION }}"
          category: 'Deployment'
          type: 'Basic'
          changelog: "https://github.com/${{ github.repository }}/blob/master/CHANGELOG.md"
          commit: "${{ github.sha }}"
          deepLink: "https://example.com/deployment"
          description: "Automated Release via Github Actions"
          user: "${{ github.actor }}"
          groupId: "deploy-group-1"

>           shortDescription: ""

Reviewer 2: Passing a blank value isn't really selling the feature. Can we just yank this line?

          timestamp: "${{ github.event.release.published_at }}"

>           customAttributes: '{cloud_vendor : "vendor_name", region : "us-east-1", environment : "staging"}'

Reviewer 2: ```suggestion
          customAttributes: '{cloud_vendor: "vendor_name", region: "us-east-1", isProd: true, instances: 2}'
```
